File: packages/flet/lib/src/controls/semantics.dart
import 'package:flutter/material.dart';

import '../extensions/control.dart';
import '../models/control.dart';
import '../utils/numbers.dart';
import '../widgets/error.dart';
import 'base_controls.dart';

class SemanticsControl extends StatelessWidget {
  final Control control;

  const SemanticsControl({super.key, required this.control});

  @override
  Widget build(BuildContext context) {
    debugPrint("Semantics build: ${control.id}");
    final content = control.buildWidget("content");
    Semantics semantics = Semantics(
      child: content ??
          const ErrorControl("Semantics.content must be provided and visible"),
      label: control.getString("label"),
      enabled: !control.disabled,
      expanded: control.getBool("expanded"),
      hidden: control.getBool("hidden"),
      selected: control.getBool("selected"),
      checked: control.getBool("checked"),
      button: control.getBool("button"),
      slider: control.getBool("slider"),
      value: control.getString("value"),
      textField: control.getBool("text_field"),
      image: control.getBool("image"),
      link: control.getBool("link"),
      header: control.getBool("header"),
      increasedValue: control.getString("increased_value"),
      decreasedValue: control.getString("decreased_value"),
      hint: control.getString("hint"),
      onTapHint: control.getString("on_tap_hint"),
      onLongPressHint: control.getString("on_long_press_hint"),
      container: control.getBool("container", false)!,
      liveRegion: control.getBool("live_region"),
      obscured: control.getBool("obscured"),
      multiline: control.getBool("multiline"),
      focused: control.getBool("focused"),
      readOnly: control.getBool("read_only"),
      focusable: control.getBool("focusable"),
      tooltip: control.getString("tooltip"),
      toggled: control.getBool("toggled"),
      maxValueLength: control.getInt("max_value_length"),
      currentValueLength: control.getInt("current_value_length"),
      headingLevel: control.getInt("heading_level"),
      excludeSemantics: control.getBool("exclude_semantics", false)!,
      mixed: control.getBool("mixed"),
      onTap: control.hasEventHandler("click")
          ? () => control.triggerEvent("click")
          : null,
      onIncrease: control.hasEventHandler("increase")
          ? () => control.triggerEvent("increase")
          : null,
      onDecrease: control.hasEventHandler("decrease")
          ? () => control.triggerEvent("decrease")
          : null,
      onDismiss: control.hasEventHandler("dismiss")
          ? () => control.triggerEvent("dismiss")
          : null,
      onScrollLeft: control.hasEventHandler("scroll_left")
          ? () => control.triggerEvent("scroll_left")
          : null,
      onScrollRight: control.hasEventHandler("scroll_right")
          ? () => control.triggerEvent("scroll_right")
          : null,
      onScrollUp: control.hasEventHandler("scroll_up")
          ? () => control.triggerEvent("scroll_up")
          : null,
      onScrollDown: control.hasEventHandler("scroll_down")
          ? () => control.triggerEvent("scroll_down")
          : null,
      onCopy: control.hasEventHandler("copy")
          ? () => control.triggerEvent("copy")
          : null,
      onCut: control.hasEventHandler("cut")
          ? () => control.triggerEvent("cut")
          : null,
      onPaste: control.hasEventHandler("paste")
          ? () => control.triggerEvent("paste")
          : null,
      onLongPress: control.hasEventHandler("long_press")
          ? () => control.triggerEvent("long_press")
          : null,
      onMoveCursorForwardByCharacter: control.getBool(
              "on_move_cursor_forward_by_character", false)!
          ? (bool value) =>
              control.triggerEvent("move_cursor_forward_by_character", value)
          : null,
      onMoveCursorBackwardByCharacter: control.getBool(
              "on_move_cursor_backward_by_character", false)!
          ? (bool value) =>
              control.triggerEvent("move_cursor_backward_by_character", value)
          : null,
      onDidGainAccessibilityFocus:
          control.hasEventHandler("did_gain_accessibility_focus")
              ? () => control.triggerEvent("did_gain_accessibility_focus")
              : null,
      onDidLoseAccessibilityFocus:
          control.hasEventHandler("did_lose_accessibility_focus")
              ? () => control.triggerEvent("did_lose_accessibility_focus")
              : null,
      onSetText: control.hasEventHandler("set_text")
          ? (String text) => control.triggerEvent("set_text", text)
          : null,
    );

    return LayoutControl(control: control, child: semantics);
  }
}

File: sdk/python/packages/flet/src/flet/controls/core/semantics.py
from typing import Optional

from flet.controls.base_control import control
from flet.controls.control import Control
from flet.controls.control_event import ControlEventHandler
from flet.controls.material.badge import BadgeValue
from flet.controls.types import Number

__all__ = ["Semantics"]


@control("Semantics")
class Semantics(Control):
    """
    Provides semantic annotations for the control tree, describing the meaning
    and purpose of controls.

    These annotations are utilized by accessibility tools, search engines,
    and semantic analysis software to better understand the structure and
    functionality of the application.
    """

    content: Optional[Control] = None
    """
    The Control to annotate.
    """

    label: Optional[str] = None
    """
    A textual description of the [`content`][(c).].
    """

    expanded: Optional[bool] = None
    """
    Whether this subtree represents something that can be in an "expanded" or
    "collapsed" state.
    """

    hidden: Optional[bool] = None
    """
    Whether this subtree is currently hidden.
    """

    selected: Optional[bool] = None
    """
    Whether this subtree represents something that can be in a selected or unselected
    state, and what its current state is.
    """

    button: Optional[bool] = None
    """
    Whether this subtree represents a button.
    """

    obscured: Optional[bool] = None
    """
    Whether [`value`][(c).] should be obscured.
    """

    multiline: Optional[bool] = None
    """
    Whether the [`value`][(c).] is coming from a field that supports multiline
    text editing.
    """

    focusable: Optional[bool] = None
    """
    Whether the node is able to hold input focus.
    """

    read_only: Optional[bool] = None
    """
    Whether this subtree is read only.
    """

    focus: Optional[bool] = None
    """
    Whether the node currently holds input focus.
    """

    slider: Optional[bool] = None
    """
    Whether this subtree represents a slider.
    """

    tooltip: Optional[str] = None
    """
    A textual description of the widget's tooltip.
    """

    badge: Optional[BadgeValue] = None
    """
    TBD
    """

    toggled: Optional[bool] = None
    """
    Whether this subtree represents a toggle switch or similar widget with an "on"
    state, and what its current state is.
    """

    max_value_length: Optional[Number] = None
    """
    The maximum number of characters that can be entered into an editable text field.
    """

    checked: Optional[bool] = None
    """
    Whether this subtree represents a checkbox or similar widget with a "checked"
    state, and what its current state is.
    """

    value: Optional[str] = None
    """
    A textual description of the `value` of the `content` control.
    """

    increased_value: Optional[str] = None
    """
    The value that the semantics node represents when it is increased.
    """

    decreased_value: Optional[str] = None
    """
    The value that the semantics node represents when it is decreased.
    """

    hint_text: Optional[str] = None
    """
    A brief textual description of the result of an action performed on the `content`
    control.
    """

    on_tap_hint_text: Optional[str] = None
    """
    TBD
    """

    current_value_length: Optional[int] = None
    """
    The current number of characters that have been entered into an editable text
    field.
    """

    heading_level: Optional[int] = None
    """
    The heading level in the DOM document structure.
    """

    exclude_semantics: bool = False
    """
    TBD
    """

    mixed: Optional[bool] = None
    """
    Whether this subtree represents a checkbox or similar control with a "half-checked"
    state or similar, and whether it is currently in this half-checked state.
    """

    on_long_press_hint_text: Optional[str] = None
    """
    TBD
    """

    container: Optional[bool] = None
    """
    TBD
    """

    live_region: Optional[bool] = None
    """
    Whether this subtree should be considered a live region.
    """

    textfield: Optional[bool] = None
    """
    Whether  this subtree represents a text field.
    """

    link: Optional[bool] = None
    """
    Whether this subtree represents a link.
    """

    header: Optional[bool] = None
    """
    Whether this subtree represents a header.
    """

    image: Optional[bool] = None
    """
    Whether the node represents an image.
    """

    on_tap: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when this control is tapped.
    """

    on_double_tap: Optional[ControlEventHandler["Semantics"]] = None
    """
    TBD
    """

    on_increase: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the value represented by the semantics node is increased.
    """

    on_decrease: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the value represented by the semantics node is decreased.
    """

    on_dismiss: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the node is dismissed.
    """

    on_scroll_left: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when a user moves their finger across the screen from right to left.
    """

    on_scroll_right: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when a user moves their finger across the screen from left to right.
    """

    on_scroll_up: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when a user moves their finger across the screen from bottom to top.
    """

    on_scroll_down: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when a user moves their finger across the screen from top to bottom.
    """

    on_copy: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the current selection is copied to the clipboard.
    """

    on_cut: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the current selection is cut to the clipboard.
    """

    on_paste: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the current content of the clipboard is pasted.
    """

    on_long_press: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the node is long-pressed (pressing and holding the screen with the
    finger for a few seconds without moving it).
    """

    on_move_cursor_forward_by_character: Optional[ControlEventHandler["Semantics"]] = (
        None
    )
    """
    Called when the cursor is moved forward by one character.
    """

    on_move_cursor_backward_by_character: Optional[ControlEventHandler["Semantics"]] = (
        None
    )
    """
    Called when the cursor is moved backward by one character.
    """

    on_did_gain_accessibility_focus: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the node has gained accessibility focus.
    """

    on_did_lose_accessibility_focus: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when the node has lost accessibility focus.
    """

    on_set_text: Optional[ControlEventHandler["Semantics"]] = None
    """
    Called when a user wants to replace the current text in the text field with a new
    text.

    Voice access users can trigger this handler by speaking type `<text>` to their
    Android devices.
    """
